File: ios/Classes/VoipLinphoneSdkPlugin.swift
import Flutter
import AVFAudio
import UIKit
import PushKit
import CallKit

public class VoipLinphoneSdkPlugin: NSObject, FlutterPlugin, FlutterStreamHandler {
    private var sipManager: SipManager = SipManager.instance
    static var eventSink: FlutterEventSink?
    private var voipRegistry: PKPushRegistry?
    
    public static func register(with registrar: FlutterPluginRegistrar) {
        let instance = VoipLinphoneSdkPlugin()
        let channel = FlutterMethodChannel(name: "voip_linphone_sdk", binaryMessenger: registrar.messenger())
        registrar.addMethodCallDelegate(instance, channel: channel)
        
        let eventChannel = FlutterEventChannel(name: "voip_linphone_sdk_event_channel", binaryMessenger: registrar.messenger())
        eventChannel.setStreamHandler(instance)
    }
    
    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        switch(call.method) {
        case "initSipModule":
            if let arguments = call.arguments as? [String:Any],
               let jsonString = arguments["sipConfiguration"] as? [String:Any],
               let jsonData = toJson(from: jsonString),
               let sipConfiguration = SipConfiguaration.toObject(JSONString: jsonData){
                do {
                    try sipManager.initSipModule(sipConfiguration: sipConfiguration)
                    result(true)
                } catch(let error) {
                    NSLog(error.localizedDescription)
                    result(FlutterError(code: "500", message: error.localizedDescription, details: nil))
                }
            } else {
                result(FlutterError(code: "500", message: "Sip configuration is not valid", details: nil))
            }
            break
        case "call":
            if let arguments = call.arguments as? [String:Any], let phoneNumber = arguments["recipient"] as? String, let isRecording = arguments["isRecording"] as? Bool {
                sipManager.call(recipient: phoneNumber, isRecording: isRecording, result: result)
            } else {
                result(FlutterError(code: "404", message: "Recipient is not valid", details: nil))
            }
            break
        case "hangup":
            sipManager.hangup(result: result)
            break
        case "answer":
            sipManager.answer(result: result)
            break
        case "reject":
            sipManager.reject(result: result)
            break
        case "transfer":
            if let arguments = call.arguments as? [String:Any], let ext = arguments["extension"] as? String {
                sipManager.transfer(recipient: ext, result: result)
            } else {
                result(FlutterError(code: "404", message: "Extension is not valid", details: nil))
            }
            
            break
        case "pause":
            sipManager.pause(result: result)
            break
        case "resume":
            sipManager.resume(result: result)
            break
        case "sendDTMF":
            if let arguments = call.arguments as? [String:Any], let dtmf = arguments["recipient"] as? String {
                sipManager.sendDTMF(dtmf: dtmf, result: result)
            } else {
                result(FlutterError(code: "404", message: "DTMF is not valid", details: nil))
            }
            
            break
        case "toggleSpeaker":
            if let arguments = call.arguments as? [String:Any], let kind = arguments["kind"] as? String {
                sipManager.toggleSpeaker(kind: kind, result: result)
            } else {
                result(FlutterError(code: "404", message: "Audio Device Kind is not valid", details: nil))
            }
            break
        case "toggleMic":
            sipManager.toggleMic(result: result)
            break
        case "refreshSipAccount":
            sipManager.refreshSipAccount(result: result)
            break
        case "unregisterSipAccount":
            sipManager.unregisterSipAccount(result: result)
            break
        case "getCallId":
            sipManager.getCallId(result: result)
            break
        case "getMissedCalls":
            sipManager.getMissCalls(result: result)
            break
        case "getSipRegistrationState":
            sipManager.getSipReistrationState(result: result)
            break
        case "isMicEnabled":
            sipManager.isMicEnabled(result: result)
            break
        case "isSpeakerEnabled":
            sipManager.isSpeakerEnabled(result: result)
            break
        case "removeListener":
            sipManager.removeListener()
            break
        case "getPlatformVersion":
            result("iOS " + UIDevice.current.systemVersion)
            break
        case "registerPush":
            registerPush()
            result(true)
            break
        case "audioDevices":
            sipManager.getAudioDevices(result: result)
            break
        case "currentAudioDevice":
            sipManager.getCurrentAudioDevice(result: result)
            break
        case "voipToken":
            let voipToken = UserDefaults.standard.string(forKey: "voipToken") ?? ""
            result(voipToken)
            break
        default:
            result(FlutterMethodNotImplemented)
        }
    }
    
    public func onListen(withArguments arguments: Any?, eventSink events: @escaping FlutterEventSink) -> FlutterError? {
        VoipLinphoneSdkPlugin.eventSink = events
        NSLog("onListen")
        return nil
    }
    
    public func onCancel(withArguments arguments: Any?) -> FlutterError? {
        VoipLinphoneSdkPlugin.eventSink = nil
        NSLog("onCancel")
        return nil
    }
    
    private func requestNotificationAuthorization() {
        UNUserNotificationCenter.current()
            .requestAuthorization(options: [.alert, .sound, .badge]) { granted, _  in
                print(">> requestNotificationAuthorization granted: \(granted)")
            }
    }
    
    private func registerPush() {
        voipRegistry = PKPushRegistry(queue: nil)
        voipRegistry?.delegate = self
        voipRegistry?.desiredPushTypes = [.voIP]
    }
}

extension VoipLinphoneSdkPlugin: UNUserNotificationCenterDelegate {
    
    public func userNotificationCenter(
        _ center: UNUserNotificationCenter,
        willPresent notification: UNNotification,
        withCompletionHandler completionHandler: (UNNotificationPresentationOptions) -> Void
    ) {
        print(">> willPresent: \(notification)")
        completionHandler([.alert, .sound, .badge])
    }
}

extension VoipLinphoneSdkPlugin: PKPushRegistryDelegate {
    
    public func pushRegistry(_ registry: PKPushRegistry, didUpdate pushCredentials: PKPushCredentials, for type: PKPushType) {
        if type == .voIP {
            let voipToken = registry.pushToken(for: .voIP)?.map { String(format: "%02X", $0) }.joined() ?? ""
            UserDefaults.standard.set(voipToken, forKey: "voipToken")
            let data = ["event": SipEvent.PushToken.rawValue, "body": ["voip_token": voipToken]] as [String: Any]
            registry.pushToken(for: .voIP)
            Self.eventSink?(data)
            
            if let rawPointer = voipToken.toUnsafeMutableRawPointer() {
                let stringLength = strlen(rawPointer.assumingMemoryBound(to: CChar.self))
                print("String length from toUnsafeMutableRawPointer(): \(stringLength)")
                let restoredString = String(cString: rawPointer.assumingMemoryBound(to: CChar.self))
                print("Restored string from toUnsafeMutableRawPointer(): \(restoredString)")
            } else {
                print("Failed to get raw pointer.")
            }
        }
    }
    
    /*public func pushRegistry(_ registry: PKPushRegistry, didReceiveIncomingPushWith payload: PKPushPayload, for type: PKPushType, completion: @escaping () -> Void) {
        if type == .voIP {
            let uuid = UUID()
            print(payload)
            
            
            //let caller = payload.dictionaryPayload["from_number"] as? String ?? ""
            //let callee = payload.dictionaryPayload["to_number"] as? String ?? ""
            //let data = ["event": SipEvent.PushReceive.rawValue, "body": ["call_id": "\(uuid)", "from_number": caller, "callee": callee]] as [String: Any]
           //Self.eventSink?(data)
            
            //reportIncommingCall(uuid, "Test", completion: completion)
        }
    }*/
    
    /*public func pushRegistry(_ registry: PKPushRegistry, didInvalidatePushTokenFor type: PKPushType) {
        
    }*/
}

extension String {
    func toUnsafeMutableRawPointer() -> UnsafeMutableRawPointer? {
        var utf8 = self.utf8CString
        return utf8.withUnsafeMutableBytes { buffer in
            if let baseAddress = buffer.baseAddress {
                return UnsafeMutableRawPointer(baseAddress.assumingMemoryBound(to: Int8.self))
            } else {
                return nil
            }
        }
    }
}

File: android/src/main/kotlin/com/voip/linphone/sdk/VoipLinphoneSdkPlugin.kt
package com.voip.linphone.sdk

import android.content.Context
import android.os.Build
import com.voip.linphone.sdk.models.SipConfiguaration
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result
import kotlinx.coroutines.CoroutineName
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.GlobalScope
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.launch

/** VoipLinphoneSdkPlugin */
class VoipLinphoneSdkPlugin : FlutterPlugin, MethodCallHandler, EventChannel.StreamHandler {
    private lateinit var channel: MethodChannel
    private lateinit var eventChannel: EventChannel
    private val sipManager: SipManager
        get() = SipManager.instance()

    private val pluginScope =
        CoroutineScope(Dispatchers.IO + SupervisorJob() + CoroutineName("VoipLinphoneSdkPluginScope"))

    private var context: Context? = null

    override fun onAttachedToEngine(flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
        channel = MethodChannel(flutterPluginBinding.binaryMessenger, "voip_linphone_sdk")
        channel.setMethodCallHandler(this)

        eventChannel =
            EventChannel(flutterPluginBinding.binaryMessenger, "voip_linphone_sdk_event_channel")
        eventChannel.setStreamHandler(this)

        context = flutterPluginBinding.applicationContext
    }

    override fun onMethodCall(call: MethodCall, result: Result) {
        when (call.method) {
            "initSipModule" -> {
                call.argument<Map<String, Any>>("sipConfiguration")?.let { arguments ->
                    val sipConfiguaration = SipConfiguaration(
                        ext = arguments["extension"] as String,
                        password = arguments["password"] as String,
                        domain = arguments["domain"] as String,
                        port = arguments["port"] as Int,
                        transportType = arguments["transportType"] as String,
                        isKeepAlive = arguments["isKeepAlive"] as Boolean
                    )

                    context?.let {
                        pluginScope.launch {
                            try {
                                sipManager.initialize(it, sipConfiguaration)
                                launch(Dispatchers.Main) {
                                    result.success(true)
                                }
                            } catch (exception: Exception) {
                                launch(Dispatchers.Main) {
                                    result.error("500", exception.localizedMessage, null)
                                }
                            }
                        }
                    } ?: run {
                        result.error("500", "Sip configuration is not valid", null)
                    }
                } ?: run {
                    result.error("500", "Sip configuration is not valid", null)
                }
            }

            "call" -> {
                val phoneNumber = call.argument<String>("recipient")
                val isRecording = call.argument<Boolean>("isRecording")

                if (phoneNumber == null || isRecording == null) {
                    result.error("404", "Recipient is not valid", null)
                } else {
                    pluginScope.launch {
                        try {
                            sipManager.call(
                                recipient = phoneNumber,
                                isRecording = isRecording,
                            )
                        } catch (exception: Exception) {
                            launch(Dispatchers.Main) {
                                result.error("500", exception.localizedMessage, null)
                            }
                        }
                    }

                }
            }

            "hangup" -> {
                pluginScope.launch {
                    try {
                        val resultHangup = sipManager.hangup()
                        launch(Dispatchers.Main) {
                            result.success(resultHangup)
                        }
                    } catch (exception: Exception) {
                        launch(Dispatchers.Main) {
                            result.error("500", exception.localizedMessage, null)
                        }
                    }
                }
            }

            "answer" -> {
                pluginScope.launch {
                    try {
                        val resultAnswer = sipManager.answer()
                        launch(Dispatchers.Main) {
                            result.success(resultAnswer)
                        }
                    } catch (exception: Exception) {
                        launch(Dispatchers.Main) {
                            result.error("500", exception.localizedMessage, null)
                        }
                    }
                }
            }

            "reject" -> {
                pluginScope.launch {
                    try {
                        val resultReject = sipManager.reject()
                        launch(Dispatchers.Main) {
                            result.success(resultReject)
                        }
                    } catch (exception: Exception) {
                        launch(Dispatchers.Main) {
                            result.error("500", exception.localizedMessage, null)
                        }
                    }
                }
            }

            "transfer" -> {
                call.argument<String>("extension")?.let { ext ->
                    pluginScope.launch {
                        try {
                            val resultTransfer = sipManager.transfer(ext)
                            launch(Dispatchers.Main) {
                                result.success(resultTransfer)
                            }
                        } catch (exception: Exception) {
                            launch(Dispatchers.Main) {
                                result.error("500", exception.localizedMessage, null)
                            }
                        }
                    }
                } ?: run {
                    result.error("404", "Extension is not valid", null)
                }
            }

            "pause" -> {
                pluginScope.launch {
                    try {
                        val resultPause = sipManager.pause()
                        launch(Dispatchers.Main) {
                            result.success(resultPause)
                        }
                    } catch (exception: Exception) {
                        launch(Dispatchers.Main) {
                            result.error("500", exception.localizedMessage, null)
                        }
                    }
                }
            }

            "resume" -> {
                pluginScope.launch {
                    try {
                        val resultResume = sipManager.resume()
                        launch(Dispatchers.Main) {
                            result.success(resultResume)
                        }
                    } catch (exception: Exception) {
                        launch(Dispatchers.Main) {
                            result.error("500", exception.localizedMessage, null)
                        }
                    }
                }
            }

            "sendDTMF" -> {
                call.argument<String>("recipient")?.let { dtmf ->
                    pluginScope.launch {
                        try {
                            val resultSend = sipManager.sendDTMF(dtmf)
                            launch(Dispatchers.Main) {
                                result.success(resultSend)
                            }
                        } catch (exception: Exception) {
                            launch(Dispatchers.Main) {
                                result.error("500", exception.localizedMessage, null)
                            }
                        }
                    }
                } ?: run {
                    result.error("404", "DTMF is not valid", null)
                }
            }

            "toggleSpeaker" -> {
                call.argument<String>("kind")?.let { kind ->
                    pluginScope.launch {
                        try {
                            val resultToggle = sipManager.toggleSpeaker(kind)
                            launch(Dispatchers.Main) {
                                result.success(resultToggle)
                            }
                        } catch (exception: Exception) {
                            launch(Dispatchers.Main) {
                                result.error("500", exception.localizedMessage, null)
                            }
                        }
                    }
                } ?: run {
                    result.error("404", "Audio Device Kind is not valid", null)
                }
            }

            "toggleMic" -> {
                pluginScope.launch {
                    try {
                        val isMicEnabled = sipManager.toggleMic()
                        launch(Dispatchers.Main) {
                            result.success(isMicEnabled)
                        }
                    } catch (exception: Exception) {
                        launch(Dispatchers.Main) {
                            result.error("500", exception.localizedMessage, null)
                        }
                    }
                }
            }

            "refreshSipAccount" -> {
                pluginScope.launch {
                    try {
                        val resultRefresh = sipManager.refreshSipAccount()
                        launch(Dispatchers.Main) {
                            result.success(resultRefresh)
                        }
                    } catch (exception: Exception) {
                        launch(Dispatchers.Main) {
                            result.error("500", exception.localizedMessage, null)
                        }
                    }
                }
            }

            "unregisterSipAccount" -> {
                pluginScope.launch {
                    try {
                        val resultUnregister = sipManager.unregisterSipAccount()
                        launch(Dispatchers.Main) {
                            result.success(resultUnregister)
                        }
                    } catch (exception: Exception) {
                        launch(Dispatchers.Main) {
                            result.error("500", exception.localizedMessage, null)
                        }
                    }
                }
            }

            "getCallId" -> {
                pluginScope.launch {
                    try {
                        val callId = sipManager.getCallId()
                        launch(Dispatchers.Main) {
                            result.success(callId)
                        }
                    } catch (exception: Exception) {
                        launch(Dispatchers.Main) {
                            result.error("500", exception.localizedMessage, null)
                        }
                    }
                }
            }

            "getMissedCalls" -> {
                pluginScope.launch {
                    try {
                        val resultMissCount = sipManager.getMissCalls()
                        launch(Dispatchers.Main) {
                            result.success(resultMissCount)
                        }
                    } catch (exception: Exception) {
                        launch(Dispatchers.Main) {
                            result.error("500", exception.localizedMessage, null)
                        }
                    }
                }
            }

            "getSipRegistrationState" -> {
                pluginScope.launch {
                    try {
                        val resultState = sipManager.getSipReistrationState()
                        launch(Dispatchers.Main) {
                            result.success(resultState)
                        }
                    } catch (exception: Exception) {
                        launch(Dispatchers.Main) {
                            result.error("500", exception.localizedMessage, null)
                        }
                    }
                }
            }

            "isMicEnabled" -> {
                pluginScope.launch {
                    try {
                        val isMicEnabled = sipManager.isMicEnabled()
                        launch(Dispatchers.Main) {
                            result.success(isMicEnabled)
                        }
                    } catch (exception: Exception) {
                        launch(Dispatchers.Main) {
                            result.error("500", exception.localizedMessage, null)
                        }
                    }
                }
            }

            "isSpeakerEnabled" -> {
                pluginScope.launch {
                    try {
                        val isSpeakerEnabled = sipManager.isSpeakerEnabled()
                        launch(Dispatchers.Main) {
                            result.success(isSpeakerEnabled)
                        }
                    } catch (exception: Exception) {
                        launch(Dispatchers.Main) {
                            result.error("500", exception.localizedMessage, null)
                        }
                    }
                }
            }

            "removeListener" -> {
                pluginScope.launch {
                    sipManager.removeListener()
                }
            }

            "getPlatformVersion" -> {
                result.success("Android ${Build.DEVICE}")
            }

            "registerPush" -> {
                result.success(true)
            }

            "audioDevices" -> {
                pluginScope.launch(Dispatchers.IO) {
                    try {
                        val resultAudioDevices = sipManager.getAudioDevices()
                        launch(Dispatchers.Main) {
                            result.success(resultAudioDevices)
                        }
                    } catch (exception: Exception) {
                        launch(Dispatchers.Main) {
                            result.error("500", exception.localizedMessage, null)
                        }
                    }
                }
            }

            "currentAudioDevice" -> {
                pluginScope.launch {
                    try {
                        val resultAudioDevice = sipManager.getCurrentAudioDevice()
                        launch(Dispatchers.Main) {
                            result.success(resultAudioDevice)
                        }
                    } catch (exception: Exception) {
                        launch(Dispatchers.Main) {
                            result.error("500", exception.localizedMessage, null)
                        }
                    }
                }
            }

            "voipToken" -> {
                result.notImplemented()
            }

            else -> {
                result.notImplemented()
            }
        }
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        channel.setMethodCallHandler(null)
        eventChannel.setStreamHandler(null)
        context = null
    }

    override fun onListen(arguments: Any?, events: EventChannel.EventSink?) {
        eventSink = events
    }

    override fun onCancel(arguments: Any?) {
        eventSink = null
    }

    companion object {
        var eventSink: EventChannel.EventSink? = null

        fun sendEvent(data: Map<String, Any?>) {
            GlobalScope.launch(Dispatchers.Main) {
                eventSink?.success(data)
            }
        }
    }
}
